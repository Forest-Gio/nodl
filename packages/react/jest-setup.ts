import '@testing-library/jest-dom';
import { matchers } from '@emotion/jest';

expect.extend(matchers);

Element.prototype.scrollTo = jest.fn();
