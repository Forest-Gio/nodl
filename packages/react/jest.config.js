module.exports = {
    testEnvironment: 'jsdom',
    snapshotSerializers: ['@emotion/jest/serializer'],
    moduleNameMapper: {
        '^(\\.{1,2}/.*)\\.js$': '$1'
    },
    transform: {
        '^.+\\.tsx?$': ['ts-jest']
    },
    globals: {
        'ts-jest': {
            tsConfig: 'tsconfig.json'
        }
    },
    setupFilesAfterEnv: ['<rootDir>/jest-setup.ts']
};
