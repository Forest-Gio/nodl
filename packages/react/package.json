{
    "name": "@nodl/react",
    "version": "0.1.13",
    "description": "React implementation of the Nodl framework",
    "main": "build/index.js",
    "types": "build/index.d.ts",
    "author": "Emil Widlund",
    "license": "MIT",
    "private": false,
    "files": [
        "build/*"
    ],
    "scripts": {
        "build": "webpack --mode=development --devtool=inline-source-map",
        "build:prod": "webpack --mode=production --devtool=source-map",
        "prepare": "yarn run build:prod",
        "test": "jest"
    },
    "repository": {
        "type": "git",
        "url": "git+https://github.com/emilwidlund/nodl.git"
    },
    "keywords": [],
    "bugs": {
        "url": "https://github.com/emilwidlund/nodl/issues"
    },
    "homepage": "https://github.com/emilwidlund/nodl#readme",
    "dependencies": {
        "@emotion/react": "^11.10.6",
        "@nodl/core": "^0.1.1",
        "@types/react": "^18.0.26",
        "mobx": "^6.7.0",
        "mobx-react-lite": "^3.4.0",
        "react": "^18.2.0",
        "react-dom": "^18.2.0",
        "react-draggable": "^4.4.5",
        "typescript": "^4.9.4"
    },
    "devDependencies": {
        "@emotion/jest": "^11.10.5",
        "@testing-library/jest-dom": "^5.16.5",
        "@testing-library/react": "^14.0.0",
        "@types/jest": "^29.2.5",
        "@types/react": "^18.0.28",
        "@types/react-dom": "^18.0.11",
        "@types/react-test-renderer": "^18.0.0",
        "@types/uuid": "^9.0.0",
        "copy-webpack-plugin": "^11.0.0",
        "jest": "^29.3.1",
        "jest-environment-jsdom": "^29.4.3",
        "ts-jest": "^29.0.3",
        "ts-loader": "^9.4.2",
        "webpack": "^5.75.0",
        "webpack-cli": "^5.0.1"
    },
    "peerDependencies": {
        "react": "^18.0.0"
    }
}
