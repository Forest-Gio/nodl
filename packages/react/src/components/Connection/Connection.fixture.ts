import { Input, Output, Node, schema } from '@nodl/core';
import { combineLatest, map } from 'rxjs';
import { z } from 'zod';

export const NumberSchema = z.number();

export class Addition extends Node {
    inputs = {
        a: new Input({ name: 'A', type: schema(NumberSchema), defaultValue: 0 }),
        b: new Input({ name: 'B', type: schema(NumberSchema), defaultValue: 0 })
    };

    outputs = {
        output: new Output({
            name: 'Output',
            type: schema(NumberSchema),
            observable: combineLatest([this.inputs.a, this.inputs.b]).pipe(
                map(inputs => inputs.reduce((sum, value) => sum + value), 0)
            )
        })
    };
}
