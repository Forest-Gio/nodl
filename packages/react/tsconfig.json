{
    "compilerOptions": {
        "lib": ["es2021", "DOM"],
        "target": "ES6",
        "strict": true,
        "moduleResolution": "node",
        "esModuleInterop": true,
        "skipLibCheck": true,
        "downlevelIteration": true,
        "outDir": "build",
        "declaration": true,
        "jsx": "react"
    },
    "include": ["./src", "./jest-setup.ts"],
    "exclude": ["./src/**/*.spec.ts", "./src/**/*.test.ts"]
}
