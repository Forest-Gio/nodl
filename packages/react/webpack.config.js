/*eslint-env node*/
const path = require('path');

module.exports = env => {
    return {
        entry: './src/index.ts',
        output: {
            filename: 'index.js',
            path: path.resolve(__dirname, 'build'),
            libraryTarget: 'umd',
            globalObject: 'this'
        },
        resolve: {
            extensions: ['.ts', '.tsx', '.js', '.json'],
            fallback: {
                stream: false,
                buffer: false
            }
        },
        mode: 'development',
        externals: {
            react: 'commonjs react',
            'react-dom': 'commonjs react-dom'
        },
        target: 'node',
        module: {
            rules: [
                {
                    test: /\.tsx?$/,
                    loader: 'ts-loader'
                }
            ]
        }
    };
};
