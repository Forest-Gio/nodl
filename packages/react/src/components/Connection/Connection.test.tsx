import { Connection } from '@nodl/core';
import { act, render, fireEvent } from '@testing-library/react';
import * as React from 'react';

import { CircuitStore, StoreContext } from '../../stores/CircuitStore/CircuitStore';
import { Connection as ConnectionComponent } from './Connection';
import { Addition } from './Connection.fixture';
import { ConnectionProps } from './Connection.types';

const setup = (store: CircuitStore, props?: Partial<ConnectionProps<unknown>>) => {
    return render(
        <StoreContext.Provider value={{ store: store }}>
            <svg width={2000} height={2000}>
                <ConnectionComponent {...props} />
            </svg>
        </StoreContext.Provider>
    );
};

describe('Connection', () => {
    let store: CircuitStore;
    let addition1: Addition;
    let addition2: Addition;
    let connection: Connection<any>;

    beforeEach(() => {
        addition1 = new Addition();
        addition2 = new Addition();

        connection = addition1.outputs.output.connect(addition2.inputs.a);

        store = new CircuitStore();

        store.setNodes([
            [addition1, { x: -400, y: 100 }],
            [addition2, { x: 400, y: 300 }]
        ]);

        // Create dummy elements so the Connection-component can compute their positions

        const node1Element = document.createElement('div');
        const node2Element = document.createElement('div');
        const outputElement = document.createElement('div');
        const inputElement = document.createElement('div');

        node1Element.style.transform = 'translate2D(-400px, 200px)';
        outputElement.style.transform = 'translate2D(-160px, 200px)';
        node2Element.style.transform = 'translate2D(400px, -200px)';
        inputElement.style.transform = 'translate2D(400px, -200px)';

        store.setNodeElement(addition1.id, node1Element);
        store.setPortElement(connection.from.id, outputElement);
        store.setNodeElement(addition2.id, node2Element);
        store.setPortElement(connection.to.id, inputElement);
    });

    it('should render correctly', () => {
        const result = setup(store, { connection });

        expect(result.asFragment()).toMatchSnapshot();
    });

    it('should dispose connection when clicked', () => {
        const result = setup(store, { connection });
        const spy = jest.spyOn(connection, 'dispose');

        expect(spy).toHaveBeenCalledTimes(0);
        expect(connection.closed).toBeFalsy();

        fireEvent.click(result.container.getElementsByTagName('path')[0]);

        expect(spy).toHaveBeenCalledTimes(1);
        expect(connection.closed).toBeTruthy();
    });

    it('should render a different color when part of selected connections', () => {
        const result = setup(store, { connection });

        expect(result.container.getElementsByClassName('connector')[0] as SVGPathElement).toHaveAttribute(
            'stroke',
            '#424763'
        );

        act(() => {
            store.selectNodes([addition1]);
        });

        const newResult = setup(store, { connection });

        expect(newResult.container.getElementsByClassName('connector')[0] as SVGPathElement).toHaveAttribute(
            'stroke',
            '#1e62ff'
        );
    });
});
