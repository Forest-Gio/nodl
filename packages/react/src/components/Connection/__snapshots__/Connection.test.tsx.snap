// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Connection should render correctly 1`] = `
<DocumentFragment>
  <svg
    height="2000"
    width="2000"
  >
    <g>
      <path
        class="connector"
        d="
        M4730,4912 
        L4974,4912 
        Q4994,4912 4994,4892 
        L4994,4732
        Q4994,4712 5014,4712
        L5258,4712
    "
        fill="none"
        stroke="#424763"
        stroke-width="2"
      />
      <path
        class="port"
        d="M4730,4912,4732,4912"
        fill="none"
        stroke="#424763"
        stroke-width="8"
      />
      <path
        class="port"
        d="M5256,4712,5258,4712"
        fill="none"
        stroke="#424763"
        stroke-width="8"
      />
    </g>
  </svg>
</DocumentFragment>
`;
